feat(std.zon): add escape_unicode options to zon.serializer

nurulhudaapon opened the pull request:
Currently std.zon.stringify.serialize always escapes Unicode characters, while std.json.stringify by default does not. This change adds an escape_unicode option that matches the JSON serializer's behavior. To maintain backward compatibility, the default value is true, preserving the current behavior of escaping Unicode.

### Change
#### Before
`std.zon.stringify.serialize(buff, .{ .whitespace = true }, writer)`
```zig
test "std.zon.stringify.serialize escape_unicode = true (default)" {
    const buff = .{ .name = "Test", .description = "⚡ Lightning Bolt", .emoji = "⚡" };
    var buff_str = std.ArrayList(u8).init(std.testing.allocator);
    defer buff_str.deinit();
    try std.zon.stringify.serialize(buff, .{ .whitespace = true }, buff_str.writer());
    std.debug.print("\n{s}\n", .{buff_str.items});
}
```
Output:
```zig
.{
    .name = "Test",
    .description = "\xe2\x9a\xa1 Lightning Bolt",
    .emoji = "\xe2\x9a\xa1",
}
```

#### After
`std.zon.stringify.serialize(buff, .{ .escape_unicode = false, .whitespace = true }, writer)`
```zig
test "std.zon.stringify.serialize escape_unicode = false (added option)" {
    const buff = .{ .name = "Test", .description = "⚡ Lightning Bolt", .emoji = "⚡" };
    var buff_str = std.ArrayList(u8).init(std.testing.allocator);
    defer buff_str.deinit();
    try std.zon.stringify.serialize(buff, .{ .escape_unicode = false, .whitespace = true }, buff_str.writer());
    std.debug.print("\n{s}\n", .{buff_str.items});
}
```
Output:
```zig
.{
    .name = "Test",
    .description = "⚡ Lightning Bolt",
    .emoji = "⚡",
}
```

### Test
```zig
const std = @import("std");

test "std.zon.stringify.serialize escape_unicode = false" {
    var buf = std.ArrayList(u8).init(std.testing.allocator);
    defer buf.deinit();

    try std.zon.stringify.serialize(
        .{ .char = "abc⚡" },
        .{ .escape_unicode = false },
        buf.writer(),
    );
    try std.testing.expectEqualStrings(".{ .char = \"abc⚡\" }", buf.items);
    buf.clearRetainingCapacity();
}
```

### Use Case

I was trying to store Unicode data in a ZON file, which I previously did in JSON. When converting from JSON to ZON using the JSON parser and ZON serializer, the Unicode characters were always escaped. This made the ZON file hard to read, which defeats its purpose as a human-readable format.

Closes https://github.com/ziglang/zig/issues/23535

nurulhudaapon committed:
feat(std.zon): add escape_unicode options to zon.serializer
Currently std.zon.stringify.serialize will always produce unicode to be escaped, whereas in std.json.stringify by default doesn't escape unicode. Adding escape_unicode option matching with the json serializer but by default it is false (as the current behaviour) to keep things backward compatible.

```zig
const std = @import("std");

test "std.zon.stringify.serialize escape_unicode = false" {
    var buf = std.ArrayList(u8).init(std.testing.allocator);
    defer buf.deinit();

    try std.zon.stringify.serialize(
        .{ .char = 'অ' },
        .{ .escape_unicode = false },
        buf.writer(),
    );
    try std.testing.expectEqualStrings(".{ .char = \"অ\" }", buf.items);
    buf.clearRetainingCapacity();
}
```

alexrp commented:
cc @MasonRemaley

MasonRemaley commented:
Thanks for the PR!

I'll take a look at this and the other Unicode related issue today. In particular, I want to look into whether or not it's necessary to maintain backwards compatibility with the current behavior.

MasonRemaley commented:
Apologies for the delay on this!

Looking it over, there was no good reason for me to escape everything by default. Adding `escape_unicode` as an option is good, and it should be false by default.

However there's one important case that needs to be addressed before this can be merged. Unless I'm missing something, the implementation here now doesn't escape `\` or `"` which is always necessary for correctness.

You can see how `std.json` handles this [here](https://github.com/ziglang/zig/blob/c06fecd46657b173763087495fe20fd720b2d210/lib/std/json/stringify.zig#L757). I *think* that's sufficient to guarantee that the output is a valid Zig string but I'll double check [stringEscape](https://ziglang.org/documentation/master/std/#std.zig.stringEscape) later to make sure there are no other important cases.

MasonRemaley commented:
Linking the issue you filed #23535 here since it's related to this PR in that it's an example of a character that can't really be printed the way you'd expect right now. We probably want to figure out how to address this as well.

nurulhudaapon committed:
wip: make escape_unicode = false by default

nurulhudaapon committed:
fix: escape ",\r,\n,\t,\

nurulhudaapon committed:
fix: make emit_codepoint_literals = .always emit double quoted hex code when needed

Previousely
`⚡` -> `'\xe2\x9a\xa1'` (Notice the hex code is single quoted which is not valid Zig/ZON syntax)

Now
`⚡` -> `"\xe2\x9a\xa1"`

`127` -> `'\x7f'` (Will still emit single quoted hex when possible)

nurulhudaapon commented:
> Apologies for the delay on this!
> 
> Looking it over, there was no good reason for me to escape everything by default. Adding `escape_unicode` as an option is good, and it should be false by default.
> 
> However there's one important case that needs to be addressed before this can be merged. Unless I'm missing something, the implementation here now doesn't escape `\` or `"` which is necessary for correctness.
> 
> You can see how `std.json` handles this [here](https://github.com/ziglang/zig/blob/c06fecd46657b173763087495fe20fd720b2d210/lib/std/json/stringify.zig#L757). I _think_ escaping these two characters is sufficient to guarantee that the output is a valid Zig string, but it's worth double checking [stringEscape](https://ziglang.org/documentation/master/std/#std.zig.stringEscape) to make sure it's not doing anything else necessary.

Updated the code to escape items that needs to be escaped following what [stringEscape](https://ziglang.org/documentation/master/std/#std.zig.stringEscape) does. I'm wondering though if it is okay to have this almost duplicated string escape logic here or just update `stringEscape` to have options to not escape unicode and re-use the same here. Otherwise everything should be good.

nurulhudaapon committed:
Merge branch 'master' into zon/serializer-unicode-escaping

nurulhudaapon commented:
Hi @MasonRemaley, let me know if the current changes look good! Thanks.

nurulhudaapon committed:
Merge branch 'master' into zon/serializer-unicode-escaping

nurulhudaapon committed:
Revert "fix: make emit_codepoint_literals = .always emit double quoted hex code when needed"

This reverts commit b9783ef85ce46186fc259ee514007ce1bd0762e5.

nurulhudaapon commented:
> This looks good, but I don't think `pub fn codePoint` is correct. As written it will sometimes output a double quoted value, and sometimes output a single quoted value. While both of these are valid outputs, they aren't actually interchangeable with eachother since one is a value and one is an array.
> 
> As you pointed out in #23535 the way I had it before wasn't correct either. One option is to just not change `codePoint` in this PR since the rest of the PR doesn't depend on it, and then merge otherwise as is. I can figure out fixing `codePoint` in a separate PR.
> 
> On the other hand, if you want to resolve #2353 in this PR, we could use the `\u{hex number}` syntax for characters that don't have recognizable escape codes like `\n`. Though maybe this should also respect the option to use escapes or not.

Thank you for reviewing. Going with the first option to not change `codePoint` here as this PR doesn't depend on it nor it is currently blocking at least my use case. Just reverted the `codePoint` changes.

Yes, `\u{hex}` should probably the best way to resolve this as it aligns with `json.stringify` as well.

nurulhudaapon committed:
Merge branch 'master' into zon/serializer-unicode-escaping

MasonRemaley reviewed (changes requested):
Looking at this a little more carefully, it's still not quite right, because some characters need to be escaped even if escapes are "off", for example, quotes.

I went ahead and fixed it locally, and also fixed #23535 by printing either \x or \u escapes as appropriate. However GitHub's "suggest changes" feature is broken for me today.

[Here's a gist with my fixes](https://gist.github.com/MasonRemaley/1bfc9d43e71d92c7b928ee77d4dcf7f4), you should be able to paste this over your file locally to see the diff.

nurulhudaapon committed:
fix: always escape some char, move escaping logic and re-use in codepoint

nurulhudaapon commented:
> Looking at this a little more carefully, it's still not quite right, because some characters need to be escaped even if escapes are "off", for example, quotes.
> 
> I went ahead and fixed it locally, and also fixed #23535 by printing either \x or \u escapes as appropriate. However GitHub's "suggest changes" feature is broken for me today.
> 
> [Here's a gist with my fixes](https://gist.github.com/MasonRemaley/1bfc9d43e71d92c7b928ee77d4dcf7f4), you should be able to paste this over your file locally to see the diff.

Thank you so much @MasonRemaley, applied the suggested changes!

nurulhudaapon committed:
Merge branch 'master' into zon/serializer-unicode-escaping

MasonRemaley commented:
No prob!

MasonRemaley reviewed (approved):
LGTM!

nurulhudaapon committed:
Merge branch 'master' into zon/serializer-unicode-escaping

nurulhudaapon committed:
Merge branch 'master' into zon/serializer-unicode-escaping

nurulhudaapon commented:
Can we get the workflow run to be approved for this PR?

nurulhudaapon committed:
Merge branch 'master' into zon/serializer-unicode-escaping

MasonRemaley commented:
@mlugg checked over this for me (I don't have power to merge) and found a couple problems with my fix.

I'll take care of these myself, but I'm listing them here for the record/so that I don't have to dig through chat to find them later:

I'll take care of them, but I'm noting them here for my own reference so I don't have to dig through chat history later:
* [ ] Right now both single and double quotes are always escaped, it would be more natural to only escape single quotes when in a single quoted literal, and double quotes when in a double quoted literal.
* [ ] Codepoint literals should always use the `\u` escape since they are unicode codepoint literals by definition, but strings should *not* since the string may not actually be unicode. Instead strings should be conservative and alway suse the `\x` escapes.
* [ ] `catch return error.InvalidCodepoint` -- this code is incorrect, surrogates (`0xD800`...`0xDFFF`)are valid codepoints and can be represented with `\u` escapes. He suggested something like [this](https://zigbin.io/f5884a).
* [ ] There are some places the code is a bit complicated in an effort to move the conditional out of the loop which may be overkill.

nurulhudaapon committed:
fix: code cleanup and single/double quote handling

nurulhudaapon committed:
fix: proper double/single quote handling and code cleanup

nurulhudaapon committed:
Merge branch 'master' into zon/serializer-unicode-escaping

nurulhudaapon committed:
refactor: rename codepoint to val for consistency

nurulhudaapon committed:
Merge branch 'master' into zon/serializer-unicode-escaping

alexrp commented:
Closing because we have no CI on GitHub anymore and this has no CI results.

Please reopen this on Codeberg: https://codeberg.org/ziglang/zig/pulls
